prover/src/logup_gkr/prover.rs

```diff
@@ -0,0 +1,256 @@
+    public_coin: &mut impl RandomCoin<BaseField = E::BaseField>,
+) -> Result<GkrCircuitProof<E>, GkrProverError> {
+    let num_logup_random_values = evaluator.get_num_rand_values();
+    let mut logup_randomness: Vec<E> = Vec::with_capacity(num_logup_random_values);
+
+    for _ in 0..num_logup_random_values {
+        logup_randomness.push(public_coin.draw().expect("failed to generate randomness"));
+    }
+
+    // evaluate the GKR fractional sum circuit
+    let circuit = EvaluatedCircuit::new(main_trace, evaluator, &logup_randomness)?;
+
+    // include the circuit output as part of the final proof
+    let CircuitLayerPolys { numerators, denominators } = circuit.output_layer().clone();
+
+    // run the GKR prover for all layers except the input layer
+    let (before_final_layer_proofs, gkr_claim) = prove_intermediate_layers(circuit, public_coin)?;
+
+    // build the MLEs of the relevant main trace columns
+    let main_trace_mls =
+        build_mls_from_main_trace_segment(evaluator.get_oracles(), main_trace.main_segment())?;
+
+    let final_layer_proof =
+        prove_input_layer(evaluator, logup_randomness, main_trace_mls, gkr_claim, public_coin)?;
+
+    Ok(GkrCircuitProof {
+        circuit_outputs: CircuitOutput { numerators, denominators },
+        before_final_layer_proofs,
+        final_layer_proof,
+    })
+}
+
+/// Proves the final GKR layer which corresponds to the input circuit layer.
+fn prove_input_layer<
+    E: FieldElement,
+    C: RandomCoin<Hasher = H, BaseField = E::BaseField>,
+    H: ElementHasher<BaseField = E::BaseField>,
+>(
+    evaluator: &impl LogUpGkrEvaluator<BaseField = E::BaseField>,
+    log_up_randomness: Vec<E>,
+    multi_linear_ext_polys: Vec<MultiLinearPoly<E>>,
+    claim: GkrClaim<E>,
+    transcript: &mut C,
+) -> Result<FinalLayerProof<E>, GkrProverError> {
+    // parse the [GkrClaim] resulting from the previous GKR layer
+    let GkrClaim { evaluation_point, claimed_evaluation } = claim;
+
+    transcript.reseed(H::hash_elements(&[claimed_evaluation.0, claimed_evaluation.1]));
+    let r_batch = transcript.draw().map_err(|_| GkrProverError::FailedToGenerateChallenge)?;
+    let claim = claimed_evaluation.0 + claimed_evaluation.1 * r_batch;
+
+    let proof = sum_check_prove_higher_degree(
+        evaluator,
+        evaluation_point,
+        claim,
+        r_batch,
+        log_up_randomness,
+        multi_linear_ext_polys,
+        transcript,
+    )?;
+
+    Ok(FinalLayerProof::new(proof))
+}
+
+// TODO: Make the multi-linears over the base field and define an operation of folding with a challenge
+// in an extension field.
+fn build_mls_from_main_trace_segment<E: FieldElement>(
+    oracles: Vec<air::LogUpGkrOracle<E::BaseField>>,
+    main_trace: &ColMatrix<<E as FieldElement>::BaseField>,
+) -> Result<Vec<MultiLinearPoly<E>>, GkrProverError> {
+    let mut mls = vec![];
+
+    for oracle in oracles {
+        match oracle {
+            air::LogUpGkrOracle::CurrentRow(index) => {
```

Comment:
```suggestion
            LogUpGkrOracle::CurrentRow(index) => {
```

Let's import `LogUpGkrOracle`

Reply:
done